### PR TITLE
Keep AppServers within range based on active sessions

### PR DESCRIPTION
**Pull request by cdonati**
This is a continuation of https://github.com/AppScale/appscale/pull/2451.

### CODE REVIEW

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/3314/

File: AppController/djinn.rb [hunk 1]
```diff
@@ -5527,31 +5539,75 @@ def get_scaling_info_for_app(app_name)
...
-    end
-
-    Djinn.log_debug("#{total_req_in_queue} requests are enqueued for app " +
-      "#{app_name} - advising that don't scale either way on this machine.")
-    return 0
+      appservers_to_scale = calculate_appservers_needed(
+        num_appengines, current_sessions, allow_concurrency)
+      Djinn.log_debug("The deployment has reached its maximum load threshold for " +
+        "app #{app_name} - Advising that we scale up #{appservers_to_scale} AppServers " +
+        "within this machine.")
```
Comment (obino) [inline AppController/djinn.rb]: I would remove the 'withing this machine'. Not sure what it means.

File: AppController/djinn.rb [hunk 2]
```diff
@@ -5527,31 +5539,75 @@ def get_scaling_info_for_app(app_name)
...
+    elsif current_load <= MIN_LOAD_THRESHOLD
+      if Time.now.to_i - @last_decision[app_name] < SCALEDOWN_THRESHOLD * DUTY_CYCLE
+        Djinn.log_debug("Not enough time has passed to scale down app #{app_name}")
+        return 0
+      end
+      appservers_to_scale = calculate_appservers_needed(
+        num_appengines, current_sessions, allow_concurrency)
+      Djinn.log_debug("The deployment is below its minimum load threshold for " +
+        "app #{app_name} - Advising that we scale down #{appservers_to_scale.abs} AppServers " +
+        "within this machine.")
```
Comment (obino) [inline AppController/djinn.rb]: Same here.

File: AppController/djinn.rb (hunk 1 above)
Comment (cdonati) [inline AppController/djinn.rb]: Good point. That's misleading.

**Comment by cdonati** [comment]
https://ocd.appscale.com:8080/job/Daily%20Build/3331/
https://ocd.appscale.com:8080/job/Hawkeye/11253/
https://ocd.appscale.com:8080/job/Tools-test/1960/
https://ocd.appscale.com:8080/job/Migration/1523/

**Pull request merged**